examples/kitt/agent.py

```diff
@@ -0,0 +1,179 @@
+    ]
+    llm_stream_queue = asyncio.Queue[LLMStream]()
+    latest_llm_stream: LLMStream = None
+    tts_stream_queue = asyncio.Queue[agents.tts.SynthesizeStream]()
+    latest_tts_stream = tts.stream()
+    working_text = ""
+    agent_speaking = False
+
+    audio_stream_future = asyncio.Future[rtc.AudioStream]()
+
+    def on_track_subscribed(track: rtc.Track, *_):
+        if track.kind == rtc.TrackKind.KIND_AUDIO:
+            audio_stream_future.set_result(rtc.AudioStream(track))
+
+    def on_chat_message(message: rtc.ChatMessage):
+        print("Got chat message: ", message)
+
+    # IMPORTANT: Subscribe to audio tracks immediately. If you don't,
+    # you are introducing a race condition where the track_publication
+    # already exists before you subscribe to it and therefore the track_subscribed
+    # event will never be fired.
+    job.room.on("track_subscribed", on_track_subscribed)
+    chat_manager.on_message(on_chat_message)
+
+    # Publish agent mic
+
+    await job.room.local_participant.publish_track(track, options)
+
+    # Wait for user audio
+    audio_stream = await audio_stream_future
+
+    async def audio_stream_task():
+        print("Starting audio stream")
+        async for audio_frame_event in audio_stream:
+            # Ignore user input while the agent is speaking
+            if agent_speaking:
+                continue
+            stt_stream.push_frame(audio_frame_event.frame)
+
+    async def stt_stream_task():
+        nonlocal working_text, latest_llm_stream, latest_tts_stream
+        async for stt_event in stt_stream:
+            # We eagerly try to run inference to keep the latency as low as possible.
+            # If we get a new transcript, we update the working text, cancel in-flight inference,
+            # and run new inference.
+            if stt_event.type == agents.stt.SpeechEventType.FINAL_TRANSCRIPT:
+                delta = stt_event.alternatives[0].text
+                # Do nothing
+                if delta == "":
+                    continue
+                working_text += delta
+                print("Working text: ", working_text)
+                working_message = ChatMessage(role=ChatRole.USER, text=working_text)
+                chat_context = agents.llm.ChatContext(
+                    messages=committed_messages + [working_message]
+                )
+                # Cancel exisiting LLM inference
+                if latest_llm_stream is not None:
+                    await latest_llm_stream.aclose(wait=False)
+                latest_llm_stream = await llm.chat(history=chat_context)
+                # Cancel existing tts inference
+                await latest_tts_stream.aclose(wait=False)
+                # Pre-warm the tts
+                latest_tts_stream = tts.stream()
+                await llm_stream_queue.put(latest_llm_stream)
+
+    async def chat_stream_task():
+        while True:
```

Comment:
trying to understand the logic here. when the user does not say a word.. `llm_stream` would be None and it would have broke out?

what happens when the user starts speaking but then stops.. would it keep waiting in `async for chunk...`?

finally, what's the significance of .flush()?